### PR TITLE
[Cluster Provisioning] Ensure changes to addon versions are reflected in addon config

### PR DESCRIPTION
**Pull request by richard-cox**
- #4714
- When provisioning an RKE2 cluster changing the kube version can not only change the addons used but also their versions (1.21 / 1.22 container network: calico).
- This change ensures addresses this by changing the way the chartValues are calculated
  - previously these were created up front and sent as a whole, though if the user changed the yaml addon config only the diff was sent
  - now only the diff is ever sent (diff is calculated on change of addon config, chartValues is then dynamically created on 'save')

Added bonuses
- user config is now retained between changes of addons (user edits config A, switches to config B, switches back to A and their edits remain). This means a user who changes addons (upgrades kube, switches container network, etc) won't lose their config after the defaults of the newly selected addon are shown.
- fixed a `Questions` related console error regarding a component with ID `Map`
- fixed an issue where `refreshYamls` could error given a missing component

Bugs
- Some addon config shows as Questions, specifically the vSphere Cloud Provider. For some headache inducing reason changes to these questions aren't retained in the input controls on blur (though are persisted where they should be

### CODE REVIEW

File: edit/provisioning.cattle.io.cluster/rke2.vue [hunk 1]
```diff
@@ -1074,6 +1097,18 @@ export default {
       return this.$store.getters['i18n/withFallback'](`cluster.addonChart."${ name }"`, null, fallback);
     },
 
+    showAddons() {
+      this.addonsRev++;
+      this.addonNames.forEach((name) => {
+        if (this.versionInfo[name]?.questions) {
+          this.userChartValuesTemp[name] = this.initYamlEditor(name);
+        } else {
+          this.userChartValuesTemp[name] = {};
```
Comment (vincent99) [inline edit/provisioning.cattle.io.cluster/rke2.vue:1106]: These need to be `set(this.userChartValuesTemp, name, …)` or the key on the object won't be reactive, which is what cause your problem with Questions disappearing on blur/rerender.

File: components/Questions/index.vue [hunk 2]
```diff
@@ -37,6 +37,8 @@ export function componentForQuestion(q) {
 
   if ( knownTypes[type] ) {
     return type;
+  } else if (type === 'map') {
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply components/Questions/index.vue:40]: I was being super defensive given the change of `knownTypes.map` --> `knownTypes.questionMap` (this fixes the `[Vue warn]: Do not use built-in or reserved HTML elements as component id: map` error given `components: { Tab, ...knownTypes },` below... which I thought might be related to the questions error but wasn't).

I'll remove then if it's always caught by `type.startsWith('map['`

File: edit/provisioning.cattle.io.cluster/rke2.vue [hunk 3]
```diff
@@ -737,11 +742,19 @@ export default {
...
+        // Allow time for addonNames to update... then fetch any missing addons
+        setTimeout(() => this.initAddons());
       }
     },
 
+    selectedVersion() {
+      this.versionInfo = {}; // Invalidate cache such that version info relevent to selected kube version is updated
+
+      // Allow time for addonNames to update... then fetch any missing addons
+      setTimeout(() => this.initAddons());
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply edit/provisioning.cattle.io.cluster/rke2.vue]: Done

File: edit/provisioning.cattle.io.cluster/rke2.vue [hunk 4]
```diff
@@ -1082,43 +1117,54 @@ export default {
...
+    syncChartValues: throttle(function(name) {
+      const fromChart = this.versionInfo[name]?.values;
+      const fromUser = this.userChartValuesTemp[name];
+      const different = diff(fromChart, fromUser);
 
-      for (const k of this.addonNames) {
-        out[k] = this.chartValues[k] || this.versionInfo[k]?.values;
+      if ( isEmpty(different) ) {
+        this.userChartValues[this.chartVersionKey(name)] = {};
+      } else {
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply edit/provisioning.cattle.io.cluster/rke2.vue]: Good point, changed

File: edit/provisioning.cattle.io.cluster/rke2.vue (hunk 1 above)
Comment (richard-cox) [reply edit/provisioning.cattle.io.cluster/rke2.vue]: Changed!

**Comment by richard-cox** [comment]
Note for QA / @sowmyav27 
The change is an area with a lot of permutation. I've tested what I can but it would be good to sanity check / smoke test the following scenarios (most importantly Harvester and vSphere)
- Create an RKE2 DO Kube version 1.21 instance with no custom addon config
- Create an RKE2 DO Kube version 1.21 instance with custom config
  - I've been using `felixConfiguration.genericXDPEnabled: true`
- Create an RKE2 DO Kube version 1.22 instance with no custom addon config
- Create an RKE2 DO Kube version 1.22 instance with custom config
  - I've been using `felixConfiguration.wireguardEnabled: true`
- Create an RKE2 Harvester Cluster, any kube version, with no custom addon config 
- Create an RKE2 Harvester Cluster, any kube version, with custom addon config
- Create an RKE2 vSphere Cluster, any kube version, with no custom addon config 
- Create an RKE2 vSphere Cluster, any kube version, with custom addon config
- Create an RKE2 AWS Cluster, any kube version, with custom addon config

Every cluster should come up correctly. Also every cluster should show the correct addon config, including anything custom, when viewing the Cluster's `Config` page `Add-on Config` tab.

Useful points...
- The combination of `Basic Tab` --> `Kubernetes Version`, `Container Network` and `Cloud Provider` dictate the add-on's and their version.
  - Kube 1.21 uses Calico  version v3.19.2 (rancher/mirrored-calico-operator v1.17.6)
  - Kube 1.22 uses Calico version v3.20.2 (rancher/mirrored-calico-operator v1.20.4)
- The configuration for each addon (Add-On Config tab) is now tied to the version of the addon
  - Switching between kube 1.21 and kube 1.22 will change the configuration shown in the Config tab
  - Any changes applied to an addon's configuration will be retained, even after removing the addon or changing it's version

**Review by vincent99 — APPROVED** [review]

**Pull request merged**